File: api/src/main/java/org/apache/flink/agents/api/Resource.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.api;

import java.util.Map;

/**
 * 抽象化的可管理资源接口
 * 
 * <p>Resource表示一个可管理的资源，包括计算资源、存储资源、网络资源、外部服务连接等。
 * 每个资源都有生命周期管理（创建、使用、释放）。
 * 
 * <p>Resource接口继承自AutoCloseable，确保资源能够被正确释放。
 * 
 * <p>示例用法：
 * <pre>{@code
 * try (Resource resource = resourceProvider.createResource(config)) {
 *     resource.initialize();
 *     if (resource.isReady()) {
 *         // 使用资源
 *         Object content = resource.getContent();
 *     }
 * }
 * }</pre>
 */
public interface Resource extends AutoCloseable {
    
    /**
     * 获取资源的唯一标识符
     * 
     * @return 资源ID
     */
    String getId();
    
    /**
     * 获取资源的类型
     * 
     * @return 资源类型
     */
    ResourceType getType();
    
    /**
     * 获取资源的当前状态
     * 
     * @return 资源状态
     */
    ResourceStatus getStatus();
    
    /**
     * 获取资源的配置属性
     * 
     * @return 资源属性映射
     */
    Map<String, Object> getProperties();
    
    /**
     * 初始化资源
     * 
     * <p>在资源可以被使用之前，必须调用此方法进行初始化。
     * 初始化过程可能包括连接建立、配置加载等操作。
     * 
     * @throws ResourceException 如果初始化失败
     */
    void initialize() throws ResourceException;
    
    /**
     * 检查资源是否准备就绪
     * 
     * <p>资源必须经过初始化并且状态为READY才能被使用。
     * 
     * @return 如果资源准备就绪返回true，否则返回false
     */
    boolean isReady();
    
    /**
     * 获取资源的实际内容
     * 
     * <p>返回资源的具体内容，类型取决于资源类型。
     * 例如，文件资源可能返回文件内容，数据库连接资源可能返回Connection对象。
     * 
     * @return 资源内容
     * @throws ResourceException 如果获取内容失败
     */
    Object getContent() throws ResourceException;
} 

File: python/flink_agents/api/resource.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
import importlib
from abc import ABC, abstractmethod
from enum import Enum
from typing import Any, Callable, Dict, Type

from pydantic import BaseModel, Field, model_serializer, model_validator


class ResourceType(Enum):
    """Type enum of resource.

    Currently, support chat_model, chat_model_server, tool, embedding_model,
    vector_store and prompt.
    """

    CHAT_MODEL = "chat_model"
    CHAT_MODEL_CONNECTION = "chat_model_connection"
    TOOL = "tool"
    EMBEDDING_MODEL = "embedding_model"
    EMBEDDING_MODEL_CONNECTION = "embedding_model_connection"
    VECTOR_STORE = "vector_store"
    PROMPT = "prompt"
    MCP_SERVER = "mcp_server"


class Resource(BaseModel, ABC):
    """Base abstract class of all kinds of resources, includes chat model,
    prompt, tools and so on.

    Resource extends BaseModel only for decreasing the complexity of attribute
    declaration of subclasses, this not represents Resource object is serializable.

    Attributes:
    ----------
    get_resource : Callable[[str, ResourceType], "Resource"]
        Get other resource object declared in the same Agent. The first argument is
        resource name and the second argument is resource type.
    """

    get_resource: Callable[[str, ResourceType], "Resource"] = Field(
        exclude=True, default=None
    )

    @classmethod
    @abstractmethod
    def resource_type(cls) -> ResourceType:
        """Return resource type of class."""


class SerializableResource(Resource, ABC):
    """Resource which is serializable."""

    @model_validator(mode="after")
    def validate_serializable(self) -> "SerializableResource":
        """Ensure resource is serializable."""
        self.model_dump_json()
        return self


class ResourceDescriptor(BaseModel):
    """Descriptor for Resource instances, storing metadata for serialization and
    instantiation.

    Attributes:
        clazz: The Python Resource class name.
        java_clazz: The Java class full path (e.g., 'com.example.YourJavaClass').
                   Empty string for Python-only resources.
        arguments: Dictionary containing resource initialization parameters.
    """
    clazz: Type[Resource] | None = None
    java_clazz: str
    arguments: Dict[str, Any]

    def __init__(self, /,
                 *,
                 clazz: Type[Resource] | None = None,
                 java_clazz: str = "",
                 **arguments: Any) -> None:
        """Initialize ResourceDescriptor.

        Args:
            clazz: The Resource class type to create a descriptor for.
            java_clazz: The Java class full path for cross-platform compatibility.
                       **REQUIRED when declaring Java resources in Python.**
                       Defaults to empty string for Python-only resources.
                       Example: "com.example.YourJavaClass"
            **arguments: Additional arguments for resource initialization.

        Usage:
            descriptor = ResourceDescriptor(clazz=YourResourceClass,
                                            param1="value1",
                                            param2="value2")
        """
        super().__init__(clazz=clazz, java_clazz=java_clazz, arguments=arguments)

    @model_serializer
    def __custom_serializer(self) -> dict[str, Any]:
        """Serialize ResourceDescriptor to dictionary.

        Returns:
            Dictionary containing python_clazz, python_module, java_clazz, and
            arguments.
        """
        return {
            "python_clazz": self.clazz.__name__,
            "python_module": self.clazz.__module__,
            "java_clazz": self.java_clazz,
            "arguments": self.arguments,
        }

    @model_validator(mode="before")
    @classmethod
    def __custom_deserialize(cls, data: dict[str, Any]) -> dict[str, Any]:
        """Deserialize data to ResourceDescriptor fields.

        Handles both new format (with python_module) and legacy format
        (full path in python_clazz).

        Args:
            data: Dictionary or other data to deserialize.

        Returns:
            Dictionary with normalized field structure.
        """
        if "clazz" in data and data["clazz"] is not None:
            return data

        args = data["arguments"]
        python_clazz = args.pop("python_clazz")
        python_module = args.pop("python_module")
        data["clazz"] = get_resource_class(python_module, python_clazz)
        data["arguments"] = args["arguments"]
        return data

    def __eq__(self, other: object) -> bool:
        """Compare ResourceDescriptor objects, ignoring private _clazz field.

        This ensures that deserialized objects (with _clazz=None) can be compared
        equal to runtime objects (with _clazz set) as long as their serializable
        fields match.
        """
        if not isinstance(other, ResourceDescriptor):
            return False
        return (
            self.clazz == other.clazz
            and self.java_clazz == other.java_clazz
            and self.arguments == other.arguments
        )

    def __hash__(self) -> int:
        """Generate hash for ResourceDescriptor, ignoring private _clazz field."""
        return hash((self.clazz, self.java_clazz,
                     tuple(sorted(self.arguments.items()))))


def get_resource_class(module_path: str, class_name: str) -> Type[Resource]:
    """Get Resource class from separate module path and class name.

    Args:
        module_path: Python module path (e.g., 'your.module.path').
        class_name: Class name (e.g., 'YourResourceClass').

    Returns:
        The Resource class type.
    """
    module = importlib.import_module(module_path)
    return getattr(module, class_name)
